File: Greedy Algorithms, Minimum Spanning Trees, and Dynamic Programming/UnionFindCompressed.rb
class UnionFindCompressed
  def initialize(objects)
    @parent = Hash.new { |h, k| h[k] = k }
    @rank = Hash.new(0)
  end

  def find(obj)
    if @parent[obj] != obj
      @parent[obj] = find(@parent[obj])
    end
    @parent[obj]
  end

  def union(obj1, obj2)
    x_root = find(obj1)
    y_root = find(obj2)
    if x_root != y_root
      if @rank[x_root] < @rank[y_root]
        @parent[x_root] = y_root
      else
        @parent[y_root] = x_root
        if @rank[x_root] == @rank[y_root]
          @rank[x_root] += 1
        end
      end
    end
  end
end

File: Greedy Algorithms, Minimum Spanning Trees, and Dynamic Programming/UnionFindCompressed.py
class UnionFindCompressed:
    def __init__(self, objects):
        self.parent = {obj: obj for obj in objects}
        self.rank = {obj: 0 for obj in objects}

    def find(self, obj):
        if self.parent[obj] != obj:
            self.parent[obj] = self.find(self.parent[obj])  # Path compression
        return self.parent[obj]

    def union(self, obj1, obj2):
        x_root = self.find(obj1)
        y_root = self.find(obj2)
        if x_root != y_root:
            if self.rank[x_root] < self.rank[y_root]:
                self.parent[x_root] = y_root
            else:
                self.parent[y_root] = x_root
                if self.rank[x_root] == self.rank[y_root]:
                    self.rank[x_root] += 1
